### PR TITLE
[Example] Add RegAE example

### PR DESCRIPTION
**Pull request by xusuyong**
### PR types

New features
### PR changes

Others
### Describe

- https://github.com/PaddlePaddle/PaddleScience/issues/440

### CODE REVIEW

File: docs/zh/examples/RegAE.md
Comment (HydrogenSulfate) [inline docs/zh/examples/RegAE.md:1]: markdown文档用 vscode markdownlint 插件格式化一下

File: examples/RegAE/RegAE.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+from ppsci.loss import KLLoss
+from ppsci.utils import logger
+
+criterion = nn.MSELoss()
+kl_loss = KLLoss()
+
+
+def loss_expr(output_dict, label_dict, weight_dict=None):
+
+    return kl_loss(output_dict) + criterion(output_dict["p_hat"], label_dict["p_hat"])
```
Comment (HydrogenSulfate) [inline examples/RegAE/RegAE.py:33]: 1. 这里用F.kl_loss和F.mse_loss吧，简化下代码
2. 函数定义放到`sup_constraint = ...` 的上方，不用作为全局函数

File: examples/RegAE/RegAE.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+    solver.train()
+    # evaluate after finished training
+    solver.eval()
+
+
+def evaluate(cfg: DictConfig):
+    # set random seed for reproducibility
+    ppsci.utils.misc.set_random_seed(cfg.seed)
+    # initialize logger
+    logger.init_logger("ppsci", osp.join(cfg.output_dir, "train.log"), "info")
```
Comment (HydrogenSulfate) [inline examples/RegAE/RegAE.py:124]: ==> `eval.log`

File: examples/RegAE/dataloader.py [hunk 3]
```diff
@@ -0,0 +1,142 @@
+"""
+输入数据类型 10^5 * 100 * 100
```
Comment (HydrogenSulfate) [inline examples/RegAE/dataloader.py:2]: 数据类型为什么是一个乘法数值？不应该是float或者double吗？还是说10^5表示数值范围，后面的100*100表示样本数？感觉表述可以更改得准确一点

File: examples/RegAE/dataloader.py [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        self.mean = 0.0
+        self.std = 1.0
+
+    def fit(self, data):
+        self.mean = np.mean(data)
+        self.std = np.std(data)
+
+    def transform(self, data):
+        mean = (
+            paddle.to_tensor(self.mean).type_as(data).to(data.device)
```
Comment (HydrogenSulfate) [inline examples/RegAE/dataloader.py:28]: ==> `to_tensor(self.mean, dtype=data.dtype)`

File: examples/RegAE/dataloader.py [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+            if paddle.is_tensor(data)
+            else self.mean
+        )
+        std = (
+            paddle.to_tensor(self.std).type_as(data).to(data.device)
+            if paddle.is_tensor(data)
+            else self.std
```
Comment (HydrogenSulfate) [inline examples/RegAE/dataloader.py:35]: 同上

File: examples/RegAE/dataloader.py [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+        )
+        return (data - mean) / std
+
+    def inverse_transform(self, data):
+        mean = paddle.to_tensor(self.mean) if paddle.is_tensor(data) else self.mean
+        std = paddle.to_tensor(self.std) if paddle.is_tensor(data) else self.std
```
Comment (HydrogenSulfate) [inline examples/RegAE/dataloader.py:41]: to_tensor(xxx, data.dtype)

File: examples/RegAE/dataloader.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+1.按照8：2划分训练数据集和测试数据集
+2.通过训练数据进行标准正则化
+"""
+import numpy as np
+import paddle
+from paddle.io import DataLoader
+from paddle.io import Dataset
+
+
+class ScalerStd(object):
```
Comment (HydrogenSulfate) [inline examples/RegAE/dataloader.py:13]: 1. 删除 `(object)`
2. 名字改为 `ZScoreNormalize` 更贴切

File: examples/RegAE/dataloader.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        )
+        return (data - mean) / std
+
+    def inverse_transform(self, data):
+        mean = paddle.to_tensor(self.mean) if paddle.is_tensor(data) else self.mean
+        std = paddle.to_tensor(self.std) if paddle.is_tensor(data) else self.std
+        return (data * std) + mean
+
+
+class ScalerMinMax(object):
```
Comment (HydrogenSulfate) [inline examples/RegAE/dataloader.py:45]: 1. 同上删除object，名字改为MinMaxNormalize

File: examples/RegAE/dataloader.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+import numpy as np
+import paddle
+from paddle.io import DataLoader
+from paddle.io import Dataset
+
+
+class ScalerStd(object):
+    """
+    Desc: Normalization utilities with std mean
+    """
```
Comment (HydrogenSulfate) [inline examples/RegAE/dataloader.py:16]: 细化docstirng，注意格式规范

File: examples/RegAE/dataloader.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+    def inverse_transform(self, data):
+        mean = paddle.to_tensor(self.mean) if paddle.is_tensor(data) else self.mean
+        std = paddle.to_tensor(self.std) if paddle.is_tensor(data) else self.std
+        return (data * std) + mean
+
+
+class ScalerMinMax(object):
+    """
+    Desc: Normalization utilities with min max
+    """
```
Comment (HydrogenSulfate) [inline examples/RegAE/dataloader.py:48]: 同格式问题

File: examples/RegAE/dataloader.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+            else self.max
+        )
+        data = 1.0 * (data - _min) / (_max - _min)
+        return 2.0 * data - 1.0
+
+    def inverse_transform(self, data, axis=None):
+        _min = paddle.to_tensor(self.min) if paddle.is_tensor(data) else self.min
+        _max = paddle.to_tensor(self.max) if paddle.is_tensor(data) else self.max
+        data = (data + 1.0) / 2.0
+        return 1.0 * data * (_max - _min) + _min
```
Comment (HydrogenSulfate) [inline examples/RegAE/dataloader.py:76]: 问题同前一个类

File: examples/RegAE/dataloader.py [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+
+
+class CustomDataset(Dataset):
```
Comment (HydrogenSulfate) [inline examples/RegAE/dataloader.py:79]: `CustomDataset`命名太泛，改为更具体的贴切名字并给出docstring

File: examples/RegAE/dataloader.py [hunk 13]
```diff
@@ -0,0 +1,142 @@
+"""
+输入数据类型 10^5 * 100 * 100
+
+1.按照8：2划分训练数据集和测试数据集
+2.通过训练数据进行标准正则化
+"""
+import numpy as np
+import paddle
+from paddle.io import DataLoader
+from paddle.io import Dataset
```
Comment (HydrogenSulfate) [inline examples/RegAE/dataloader.py:10]: ==> `from paddle import io`，下方调用改为`io.xxx`

File: examples/RegAE/dataloader.py [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+
+        self.data_type = data_type
+
+        self.train_len = int(num * 0.8)
+        self.test_len = num - self.train_len
+
+        self.train_data = data[: self.train_len]
+        self.test_data = data[self.train_len :]
+
+        self.scaler = ScalerStd()
```
Comment (HydrogenSulfate) [inline examples/RegAE/dataloader.py:107]: scaler跟AMP的scaler存在歧义，改为 normalizer

File: ppsci/arch/__init__.py
Comment (HydrogenSulfate) [inline ppsci/arch/__init__.py:1]: API文档里加一下AutoEncoder

File: ppsci/arch/vae.py [hunk 15]
```diff
@@ -0,0 +1,74 @@
...
+
+from typing import Tuple
+
+import paddle
+import paddle.nn as nn
+
+from ppsci.arch import base
+
+
+# copy from AISTUDIO
```
Comment (HydrogenSulfate) [inline ppsci/arch/vae.py:25]: 这行删掉吧

File: ppsci/arch/vae.py [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+class AutoEncoder(base.Arch):
+    def __init__(
+        self,
+        input_keys: Tuple[str, ...],
+        output_keys: Tuple[str, ...],
+        input_dim,
+        latent_dim,
+        hidden_dim,
+    ):
```
Comment (HydrogenSulfate) [inline ppsci/arch/vae.py:34]: 完善docstring、type hint

File: ppsci/loss/kl.py [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+            )
+        super().__init__(reduction, weight)
+
+    def forward(self, output_dict, label_dict=None, weight_dict=None):
+        mu, log_sigma = output_dict["mu"], output_dict["log_sigma"]
+
+        base = paddle.exp(2.0 * log_sigma) + paddle.pow(mu, 2) - 1.0 - 2.0 * log_sigma
+        loss = 0.5 * paddle.sum(base) / mu.shape[0]
+
+        return loss
```
Comment (HydrogenSulfate) [inline ppsci/loss/kl.py:45]: 这是KL散度损失吗？但是看起来跟nn.KLDivLoss不是很像，如果不是很通用的Loss的话还是放在案例里通过FunctionalLoss使用吧，如果是比较通用的，就需要写成KLDiv的形式，即 $p*(\ln{\frac{p}{q}})$ ，并写好带有数学公式的 docstring

**Review by HydrogenSulfate — CHANGES_REQUESTED** [review]
辛苦修改一下

**Review by HydrogenSulfate — APPROVED** [review]
LGTM
